quinn-proto/src/connection.rs

```diff
@@ -512,22 +597,27 @@ impl Connection {
             }
             // Don't apply congestion penalty for lost ack-only packets
             let lost_nonack = old_bytes_in_flight != self.bytes_in_flight;
-            // Start a new recovery epoch if the lost packet is larger than the end of the
-            // previous recovery epoch.
-            if lost_nonack && !self.in_recovery(largest_lost) {
-                self.end_of_recovery = self.largest_sent_packet();
-                // *= factor
-                self.congestion_window =
-                    (self.congestion_window * self.config.loss_reduction_factor as u64) >> 16;
-                self.congestion_window =
-                    cmp::max(self.congestion_window, self.config.minimum_window);
-                self.ssthresh = self.congestion_window;
+            if lost_nonack {
+                self.congestion_event(now, largest_lost_time)
             }
         }
     }
 
-    fn in_recovery(&self, packet: u64) -> bool {
-        packet <= self.end_of_recovery
+    fn congestion_event(&mut self, now: u64, sent_time: u64) {
```

Comment:
Next time, please put the method extraction in a separate commit from the logic changes? Reviewing both at the same time is painful.